dialogue-annotations-processor/src/test/java/com/palantir/myservice/service/MyService.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.myservice.service;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.HttpMethod;
+import com.palantir.dialogue.RequestBody;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.annotations.DialogueClient;
+import com.palantir.dialogue.annotations.Request;
+import java.util.UUID;
+
+@DialogueClient
+public interface MyService {
```

Comment:
Out of curiosity, how is the interface expected to be used since its missing the static factory method typical of Dialogue interfaces? Will the client factory be aware of the annotations and the generated classes?

Reply:
We'd discussed requiring the generated class ref `value` of the `@DialogueClient` annotation, and teaching the factory to follow links. I don't think we can get away with requiring folks to pass the generated class to the client factory because it's too easy to get wrong, and fails at runtime.

Reply:
Very much open to suggestions if you have ideas :-)

Reply:
I think it would actually be kinda nice to expect users to do something like:

```java
wc.client(DialogueMyService.class, "my-service").get();
```

supposing we generated something like:

```java
interface DialogueMyService extends MyService {
  static DialogueMyService of(Channel channel, ConjureRuntime runtime) {
    ...
  }
}
```

I worry that if we allowed:
```java
wc.client(MyService.class, "my-service").get();
```

that it would get a bit confusing for users what was going on under the covers and that you could end up in a split state where sometimes the generated interface is used over the declared interface.

Reply:
Right, that's exactly what I'm worried about because the type system isn't giving us any safety, and based on all our existing clients it feels like the right thing to do.

Reply:
We'll codegen it eventually, right now just spitballing. Of course I forgot to add the "draft" PR thing... 🤦

Reply:
So if I understand it, your worries are:

1. People will sometimes pass the generated interface, and sometimes the interface they hand annotate.
2. If we pass in the hand-annotated one, it's possible they haven't actually run the generator, and so if witchcraft were to go looking for the generated class, it would not be found?

Doing:

```
MyService myService = wc.client(DialogueMyService.class, "my-service").get();
```

feels unnecessary to me and is not what we do for conjure generated ones: you pass in the DialogueClientBlocking or DialogueClientAsync right? So there's no asymmetry like is being proposed here.

Personally, I'd rather we actually make a gradle plugin or something lightweight that people apply, and that thing correctly sets up the annotation processor and the correct compileOnly dependency, so it's not like with immutables were most people forget about the compileOnly dependency and get weird warnings.

But maybe I'm misunderstanding the concern?

Reply:
If we require the generated interface to be passed to the conjure factory, forgetting to do so and passing the raw interface (which is what we do for conjure-generated interfaces) will compile locally correctly, and fail at runtime. We generally want to fail at compilation rather than runtime.

So we need a way to map from `MyService.class` to `DialogueMyService.class` safely so we can teach the generator to look for factory methods on the latter. One of the easiest ways is to include a direct reference to the generated class from our `DialogueClient` annotation, this will both show the factory which interface to construct, as well as enforce at compile time that the processor has been applied, otherwise the DialogueMyService.class cannot be located.

Alternatively we could do something clever with a new service-loadable factory of the form:
```java
interface DialogueServiceFactory<MyService> {
  MyService create(EndpointChannelFactory channel, ConjureRuntime runtime);
}
```
I like the idea of avoiding the reflection hackery we currently use, but I think this solution could be overkill.

Reply:
We can do the "include direct reference to the generated class", but I feel that a small gradle plugin that sets up dependencies correctly, and then generates the class in the same package as the annotated class works? Then the Witchcraft factory can assume that it should just load the ``"Dialogue" + clazz.getFullyQualifiedClass()``.

The custom annotation works for me, I don't have a strong preference other than the workflow of "add plugin and write your custom interface" feels less work than "correctly setup dependencies -> write interface with annotations -> annotate the interface".

Reply:
imo gradle plugins are a pain, and specific to gradle. They can break in their own odd set of ways, so I'm not convinced a plugin is the right way to tackle this problem. That takes us from setting up 2 depenedncies, to setting up 3 dependencies, one of which may set up the other two if it works properly ;-)

The more I think about this, the better I like a more generic solution that could simplify our existing conjure-generated clients as well. I think our magical static method lookups were a mistake, we don't need to further compound upon them especially given newer jres reflection restrictions.

Reply:
Alright, so:

1. DialogueMyService is generated by the annotation processor.
2. We annotate MyService with @DialogueClient(DialogueMyService.class)
3. DialogueMyService implements DialogueServiceFactory<MyService>
4. Clients use this: ``MyService myService = wc.client(DialogueMyService.class, "my-service").get();``

Did I miss something?

Reply:
1. :heavy_check_mark: DialogueMyService is generated by the annotation processor.
2. :heavy_check_mark: We annotate MyService with @DialogueClient(DialogueMyService.class)
3. :yellow_circle: DialogueMyService implements DialogueServiceFactory: This one is somewhat nebulous. Either our annotation points to a standard dialogue service, or we reimplement the way dialogue services are created. I'd like to improve how all dialogue services are created, so I'll try to prototype something this evening. Failing that, the annotaiton will point to the "concrete" dialogue interface which our client factory will follow and create.
4. :x: Clients use this: `MyService myService = wc.client(DialogueMyService.class, "my-service").get()`: This is dangerous because the type system allows folks to pass MyService.class without failing compilation. We should hide the generated code as much as possible at this point.

Reply:
For initial prototyping we can use `factory.client(DialogueMyService.class, "my-service")` and improve the system in parallel, but I wouldn't want anyone to use it that way.

Reply:
https://github.com/palantir/dialogue/pull/1138